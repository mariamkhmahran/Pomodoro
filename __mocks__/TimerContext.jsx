import context from "src/logic/TimerContext/testHelpers";

export const Context = {
  Consumer(props) {
    return props.children(context);
  }
};
