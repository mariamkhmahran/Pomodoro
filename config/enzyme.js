import { configure } from "enzyme";
import Adapter from "enzyme-react-adapter-future";

configure({ adapter: new Adapter() });
