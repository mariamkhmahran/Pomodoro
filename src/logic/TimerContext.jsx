import React from "react";

const POMODORO_TIME = 0.025 * 60 * 1000;
export const Context = React.createContext();

class TimerContext extends React.Component {
  time = POMODORO_TIME;

  tick = () => {
    const { isTimerRunning, remainigTime } = this.state;
    if (isTimerRunning) {
      let currentTime = Date.now();
      if (remainigTime >= 0) {
        this.setState({
          remainigTime: this.time - (currentTime - this.startTime)
        });
        window.requestAnimationFrame(this.tick);
      } else {
        this.completeCycle();
      }
    }
  };

  completeCycle = () => {
    this.setState(
      () => ({
        remainigTime: POMODORO_TIME,
        isTimerRunning: false,
        endOfCycle: true
      }),
      () => {
        this.setState({ endOfCycle: false });
      }
    );
  };

  startTimer = () => {
    this.setState(
      () => ({
        isTimerRunning: true,
        endOfCycle: false
      }),
      () => {
        this.startTime = Date.now();
        window.requestAnimationFrame(this.tick);
      }
    );
  };

  stopTimer = () => {
    this.time = this.state.remainigTime;
    this.setState({
      isTimerRunning: false
    });
  };

  state = {
    remainigTime: POMODORO_TIME,
    startTimer: this.startTimer,
    stopTimer: this.stopTimer,
    isTimerRunning: false,
    endOfCycle: false
  };

  render() {
    return (
      <Context.Provider value={this.state}>
        {this.props.children}
      </Context.Provider>
    );
  }
}
export default TimerContext;
