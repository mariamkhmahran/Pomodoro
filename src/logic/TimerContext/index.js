export { default } from "./TimerContext";
export * from "./TimerContext";
