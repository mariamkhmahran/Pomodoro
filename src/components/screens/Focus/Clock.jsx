import React from "react";
import moment from "moment";
import { Context } from "src/logic/TimerContext";

const Clock = () => (
  <div>
    <Context.Consumer>
      {({ remainigTime }) => (
        <h1 id="clock">{moment(remainigTime).format("mm:ss")}</h1>
      )}
    </Context.Consumer>
  </div>
);

export default Clock;
