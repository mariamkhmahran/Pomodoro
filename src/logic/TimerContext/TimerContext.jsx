import React from "react";
import * as GeneralState from "src/logic/GeneralState";

const POMODORO_TIME = 25 * 60 * 1000;
export const Context = React.createContext();

class TimerContext extends React.Component {
  time = POMODORO_TIME;

  tick = () => {
    const { isTimerRunning, remainigTime } = this.state;
    if (isTimerRunning) {
      let currentTime = new Date();
      if (remainigTime >= 0) {
        this.setState({
          remainigTime: this.time - (currentTime - this.startTime)
        });
        window.requestAnimationFrame(this.tick);
      } else {
        this.completeCycle();
      }
    }
  };

  completeCycle = () => {
    this.setState(() => ({
      remainigTime: POMODORO_TIME,
      isTimerRunning: false
    }));
    this.props.addCycle();
    GeneralState.logCycle();
  };

  startTimer = () => {
    this.setState(
      () => ({
        isTimerRunning: true
      }),
      () => {
        this.startTime = Date.now();
        window.requestAnimationFrame(this.tick);
      }
    );
  };

  stopTimer = () => {
    this.time = this.state.remainigTime;
    this.setState({
      isTimerRunning: false
    });
  };

  state = {
    remainigTime: POMODORO_TIME,
    startTimer: this.startTimer,
    stopTimer: this.stopTimer,
    isTimerRunning: false
  };

  render() {
    return (
      <Context.Provider value={this.state}>
        {this.props.children}
      </Context.Provider>
    );
  }
}
export default TimerContext;
