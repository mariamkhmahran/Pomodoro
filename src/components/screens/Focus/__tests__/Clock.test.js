import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Clock from "src/components/screens/Focus/Clock";
import TimerContext from "src/logic/TimerContext/TimerContext";

test("renders correctly", () => {
  const wrapper = renderer
    .create(
      <TimerContext>
        <Clock />
      </TimerContext>
    )
    .toJSON();

  expect(wrapper).toMatchSnapshot();
});

test("stop and start timer", () => {
  const wrapper = mount(
    <TimerContext>
      <Clock />
    </TimerContext>
  );

  wrapper
    .childAt(0)
    .find("button")
    .at(0)
    .simulate("click");

  expect(wrapper.instance().state.isTimerRunning).toEqual(true);
});
