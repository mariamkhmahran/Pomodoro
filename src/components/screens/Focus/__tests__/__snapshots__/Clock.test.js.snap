// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="left_col"
>
  <div>
    <h1
      id="clock"
    >
      25:00
    </h1>
    <button
      className="no-btn-styles round"
      onClick={[Function]}
    >
      <img
        alt="start"
        className="circular_btn"
        id="play"
        src="play_btn.png"
      />
    </button>
  </div>
</div>
`;
